### PR TITLE
Use AutoMocker to create parameters and to create Mocks

### PR DESCRIPTION
**Pull request by joelweiss**
This solves having parameters that have constructor parameters, and all the dependencies are mocked.

### CODE REVIEW

File: Moq.AutoMock/Instance.cs [hunk 1]
```diff
@@ -53,15 +54,27 @@ public MockInstance(Mock value)
...
         }
 
-        private static Mock CreateMockOf(Type type, MockBehavior mockBehavior)
+        private static Mock CreateMockOf(AutoMocker autoMocker, Type type, MockBehavior mockBehavior, BindingFlags bindingFlags)
         {
-            var mockType = typeof (Mock<>).MakeGenericType(type);
-            var mock = (Mock) Activator.CreateInstance(mockType, mockBehavior);
+            if (type.IsClass && !type.IsAbstract)
+            {
+                MethodInfo method = typeof(AutoMocker).GetMethods().Single(mi => mi.Name == nameof(AutoMocker.CreateSelfMock) && mi.GetParameters().SingleOrDefault(pi => pi.ParameterType == typeof(bool)) != null);
```
Comment (tkellogg) [inline Moq.AutoMock/Instance.cs]: This is a very long line, please shorten it

**Comment by tkellogg** [comment]
I'm not sure if I understand what you're doing. Regardless if I understand, I think we're crossing a line regarding the simplicity of the code. Please refactor to make it more obvious what's going on. Thanks

**Comment by joelweiss** [comment]
You are right, I first did it like that, to start the discussion if you want to go that route.

The current AutoMocker public API only supports a generic call to `CreateSelfMock<T>` (I understand why because the Moq API also only supports `new Mock<T>`, and it only makes sense to expose it like that) and I have to call it for types I don't know at compile time, so before changing `AutoMocker` itself I first implemented a proof of concept with reflection.

Now what I'm trying to do.
When you call `CreateSelfMock<T>` all the dependencies (parameters) was created with 
```csharp
private static Mock CreateMockOf(Type type, MockBehavior mockBehavior) 
{ 
    var mockType = typeof (Mock<>).MakeGenericType(type); 
    var mock = (Mock) Activator.CreateInstance(mockType, mockBehavior); 
    return mock; 
} 
```
but since `Moq` needs all the parameters when creating the `Mock<T>`, it will fail if any of the dependencies have other dependencies, so what I'm doing is for each parameter I use the `AutoMocker` to instantiate.
So to sum it up all dependencies of `T` and its dependencies and so on are created with `AutoMocker`.

See example of failing code.
```csharp
using System;

public class Program
{
	public static void Main()
	{
		Console.WriteLine("Start Test");
		var autoMocker = new Moq.AutoMock.AutoMocker();
		var bar = autoMocker.CreateSelfMock<Bar>();
		Console.WriteLine("So far so good");
		var foo = autoMocker.CreateSelfMock<Foo>();
                Console.WriteLine("I wish I got until here :-(");
	}

	public class Foo
	{
		public Foo(Bar bar)	{	}
	}
	
	public class Bar
	{
		public Bar(Baz baz)	{	}
	}
	
	public class Baz
	{
	}
}
```

**Comment by Alex2357** [comment]
I believe this is needed. I also did similar method in my forked version. 
Let me explain why I need this.
For complex integration tests when I need to test a complex object graph what I do is that I create my DI container(Castle Windsor), instead of some implementations I need mocked versions of some objects. Castle Windsor container gets ISubDependencyResolver which for certain types need to do requests to Automocker instance. At that point I don't have a generic type I just have a Type.

**Comment by Keboo** [comment]
Given how old this PR is, with many merge conflicts, I am closing it. I have created issue #38 to track the feature request.
